go/chat/attachment_stash.go

```diff
@@ -0,0 +1,176 @@
+}
+
+type FileStash struct {
+	sync.Mutex
+}
+
+func (f *FileStash) Start(plaintextHash []byte, conversationID chat1.ConversationID, info AttachmentInfo) error {
+	f.Lock()
+	defer f.Unlock()
+	c, err := f.contents()
+	if err != nil {
+		return err
+	}
+	info.StartedAt = time.Now()
+	key := stashKey{PlaintextHash: plaintextHash, ConversationID: conversationID}
+	c[key.String()] = info
+
+	return f.serialize(c)
+}
+
+func (f *FileStash) Lookup(plaintextHash []byte, conversationID chat1.ConversationID) (AttachmentInfo, bool, error) {
+	f.Lock()
+	defer f.Unlock()
+	return f.lookup(plaintextHash, conversationID)
+}
+
+func (f *FileStash) RecordPart(plaintextHash []byte, conversationID chat1.ConversationID, partNumber int, hash string) error {
+	f.Lock()
+	defer f.Unlock()
+	c, err := f.contents()
+	if err != nil {
+		return err
+	}
+	key := stashKey{PlaintextHash: plaintextHash, ConversationID: conversationID}
+	info, found := c[key.String()]
+	if !found {
+		return errors.New("in progress upload not found")
+	}
+
+	if info.Parts == nil {
+		info.Parts = make(map[int]string)
+	}
+
+	info.Parts[partNumber] = hash
+	c[key.String()] = info
+	return f.serialize(c)
+}
+
+func (f *FileStash) VerifyPart(plaintextHash []byte, conversationID chat1.ConversationID, partNumber int, hash string) (bool, error) {
+	f.Lock()
+	defer f.Unlock()
+	info, found, err := f.lookup(plaintextHash, conversationID)
+	if err != nil {
+		return false, err
+	}
+	if !found {
+		return false, errors.New("in progress upload not found")
+	}
+
+	return info.Parts[partNumber] == hash, nil
+}
+
+func (f *FileStash) Finish(plaintextHash []byte, conversationID chat1.ConversationID) error {
+	f.Lock()
+	defer f.Unlock()
+	c, err := f.contents()
+	if err != nil {
+		return err
+	}
+	key := stashKey{PlaintextHash: plaintextHash, ConversationID: conversationID}
+	delete(c, key.String())
+	return f.serialize(c)
+}
+
+func (f *FileStash) filename() string {
+	return filepath.Join(os.TempDir(), "chat_attachment_stash")
+}
+
+func (f *FileStash) contents() (map[string]AttachmentInfo, error) {
+	x, err := os.Open(f.filename())
+	if err != nil {
+		if os.IsNotExist(err) {
+			return make(map[string]AttachmentInfo), nil
+		}
+		return nil, err
+	}
+	defer x.Close()
+
+	v := make(map[string]AttachmentInfo)
+	dec := gob.NewDecoder(x)
+	if err := dec.Decode(&v); err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+
+func (f *FileStash) serialize(m map[string]AttachmentInfo) error {
+	x, err := os.Create(f.filename())
+	if err != nil {
+		return err
+	}
+	defer x.Close()
+	enc := gob.NewEncoder(x)
+	return enc.Encode(m)
+}
+
+func (f *FileStash) lookup(plaintextHash []byte, conversationID chat1.ConversationID) (AttachmentInfo, bool, error) {
+	c, err := f.contents()
+	if err != nil {
+		return AttachmentInfo{}, false, err
+	}
+	key := stashKey{PlaintextHash: plaintextHash, ConversationID: conversationID}
+	info, found := c[key.String()]
+	return info, found, nil
+}
```

Comment:
I expect it to be empty.  I'd rather have one source of truth (disk) than worry about memory and disk.